README
NAME
    App::AYCABTU - All Your Codes Are Belong To Us

SYNOPSIS
        > aycabtu --help

DESCRIPTION
    This module installs a program called aycabtu, that can be used to
    manage all of the code repositories that you are interested in.

STATUS
    this is a very early release. Only a couple features are implemented,
    and somewhat poorly.

    See <http://github.com/ingydotnet/aycabt-ingydotnet/blob/master/AYCABTU>
    for an example of how to configure AYCABTU.

    Stay tuned. Things should be much better soon.

RESOURCES
    CPAN: <http://search.cpan.org/dist/App-AYCABTU/>

    GitHub: <http://github.com/ingydotnet/app-aycabtu-pm>

AUTHOR
    Ingy döt Net <ingy@cpan.org>

COPYRIGHT
    Copyright 2010-2014. Ingy döt Net

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    See <http://www.perl.com/perl/misc/Artistic.html>



Changes
0.12 Sat Aug 16 13:19:52 PDT 2014
 - Some doc fixes. mst++ for reals this time.

0.11 Sat Aug 16 13:14:27 PDT 2014
 - Fix Meta homepage. mst++

0.10 Sat Aug 16 10:47:40 PDT 2014
 - Meta 0.0.2

0.09 Fri Aug 15 18:02:47 PDT 2014
 - Remove t/_

0.08 Fri Aug 15 15:39:33 PDT 2014
 - Add t/000-compile-modules.t
 - CPAN Day 2014 - 21 Release Salute!

0.07 Fri Aug  8 11:50:03 PDT 2014
 - Swith hashbangs to `/usr/bin/env perl`

0.06 Sat Aug  2 10:03:19 PDT 2014
 - Fix Copyright years.

0.05 Sat Jul 26 22:04:46 PDT 2014
 - Switch to Zilla-Dist

0.04 Mon Jun  6 00:15:51 EST 2011
 - Allow single numbers for selection
 - Switch to Module::Package and Mouse.

0.03 Thu Nov 18 22:31:39 PST 2010
 - Glob together multiple config files.

0.02 Tue Oct  5 21:30:41 PDT 2010
 - Add git clone support
 - Add names selection

0.01 Sat Sep 18 00:36:35 PDT 2010
 - Initial version.



inc/TestML/Runtime.pm
#line 1
package TestML::Runtime;
use TestML::Base -base;
use TestML::Compiler;

# Since there is only ever one test runtime, it makes things a LOT cleaner to
# keep the reference to it in a global variable accessed by a method, than to
# put a reference to it into every object that needs to access it.
our $self;

has 'base', -init => '$0 =~ m!(.*)/! ? $1 : "."';   # Base directory
has 'testml';           # TestML document filename, handle or text
has 'bridge';           # Bridge transform module
# XXX Add TestML.pm support for -library keyword.
has 'library' => [];    # Transform library modules

has 'function';         # Current function executing
has 'planned' => 0;     # plan() has been called
has 'test_number' => 0; # Number of tests run so far

sub init {
    my $self = $TestML::Runtime::self = shift;
    $self->SUPER::init(@_);
    $self->function($self->compile_testml);
    $self->load_variables;
    $self->load_transform_module('TestML::Library::Standard');
    $self->load_transform_module('TestML::Library::Debug');
    if ($self->bridge) {
        $self->load_transform_module($self->bridge);
    }
    return $self;
}

# XXX Move to TestML::Adapter
sub title { }
sub plan_begin { }
sub plan_end { }

sub run {
    my $self = shift;

    my $function = $self->function;
    my $context = TestML::None->new;
    my $args = [];

    $self->run_function($self->function, $context, $args);

    $self->run_plan();
    $self->plan_end();
}

# XXX - TestML exception handling needs to happen at the function level, not
# just at the expression level. Not yet handled here.
sub run_function {
    my $self = shift;
    my $function = shift;
    my $context = shift;
    my $args = shift;

    my $signature = $function->signature;
    die sprintf(
        "Function received %d args but expected %d",
        scalar(@$args),
        scalar(@$signature),
    ) if @$signature and @$args != @$signature;
    $function->setvar('Self', $context);
    for (my $i = 0; $i < @$signature; $i++) {
        my $arg = $args->[$i];
        $arg = $self->run_expression($arg)
            if ref($arg) eq 'TestML::Expression';
        $function->setvar($signature->[$i], $arg);
    }

    my $parent = $self->function;
    $self->function($function);

    for my $statement (@{$function->statements}) {
        $self->run_statement($statement);
    }

    $self->function($parent);

    return TestML::None->new;
}

sub run_statement {
    my $self = shift;
    my $statement = shift;
    my $blocks = @{$statement->points}
        ? $self->select_blocks($statement->points)
        : [1];
    for my $block (@$blocks) {
        $self->function->setvar('Block', $block) if ref($block);
        my $context = $self->run_expression($statement->expression);
        if (my $assertion = $statement->assertion) {
            $self->run_assertion($context, $assertion);
        }
    }
}

sub run_assertion {
    my $self = shift;
    my $left = shift;
    my $assertion = shift;
    my $method = 'assert_' . $assertion->name;

    # Run this as late as possible.
    $self->run_plan;

    $self->test_number($self->test_number + 1);
    $self->function->setvar(
        TestNumber => TestML::Num->new(value => $self->test_number),
    );

    # TODO - Should check 
    my $results = ($left->type eq 'List')
        ? $left->value
        : [ $left ];
    for my $result (@$results) {
        if (@{$assertion->expression->units}) {
            my $right = $self->run_expression($assertion->expression);
            my $matches = ($right->type eq 'List')
                ? $right->value
                : [ $right ];
            for my $match (@$matches) {
                $self->$method($result, $match);
            }
        }
        else {
            $self->$method($result);
        }
    }
}

sub run_expression {
    my $self = shift;
    my $prev_expression = $self->function->expression;
    my $expression = shift;
    $self->function->expression($expression);

    my $units = $expression->units;
    my $context = TestML::None->new;

    for (my $i = 0; $i < @$units; $i++) {
        my $unit = $units->[$i];
        if ($expression->error) {
            next unless
                $unit->isa('TestML::Transform') and
                $unit->name eq 'Catch';
        }
        if ($unit->isa('TestML::Object')) {
            $context = $unit;
            next;
        }
        if ($unit->isa('TestML::Function')) {
            $context = $unit;
            next;
        }
        die "Unexpected unit: $unit" unless $unit->isa('TestML::Transform');
        my $callable = $self->function->getvar($unit->name)
            or die "Can't find transform '${\$unit->name}'";
        my $args = $unit->args;
        if ($callable->isa('TestML::Native')) {
            $context = $self->run_native($callable->value, $context, $args);
        }
        elsif ($callable->isa('TestML::Object')) {
            $context = $callable;
        }
        elsif ($callable->isa('TestML::Function')) {
            if ($i or $unit->explicit_call) {
                my $points = $self->function->getvar('Block')->points;
                for my $key (keys %$points) {
                    $callable->setvar($key, TestML::Str->new(value => $points->{$key}));
                }
                $context = $self->run_function($callable, $context, $args);
            }
            $context = $callable;
        }
        else {
            ZZZ $expression, $unit, $callable;
        }
    }
    if ($expression->error) {
        die $expression->error;
    }
    $self->function->expression($prev_expression);
    return $context;
}

sub run_native {
    my $self = shift;
    my $function = shift;
    my $context = shift;
    my $args = shift;
    my $value = eval {
        &$function(
            $context,
            map {
                (ref($_) eq 'TestML::Expression')
                ? $self->run_expression($_)
                : $_
            } @$args
        );
    };
    if ($@) {
        $self->function->expression->error($@);
        $context = TestML::Error->new(value => $@);
    }
    elsif (UNIVERSAL::isa($value, 'TestML::Object')) {
        $context = $value;
    }
    else {
        $context = $self->object_from_native($value);
    }
    return $context;
}

sub select_blocks {
    my $self = shift;
    my $wanted = shift;
    my $selected = [];

    OUTER: for my $block (@{$self->function->data}) {
        my %points = %{$block->points};
        next if exists $points{SKIP};
        for my $point (@$wanted) {
            next OUTER unless exists $points{$point};
        }
        if (exists $points{ONLY}) {
            @$selected = ($block);
            last;
        }
        push @$selected, $block;
        last if exists $points{LAST};
    }
    return $selected;
}

sub object_from_native {
    my $self = shift;
    my $value = shift;
    return
        not(defined $value) ? TestML::None->new :
        ref($value) eq 'ARRAY' ? TestML::List->new(value => $value) :
        $value =~ /^-?\d+$/ ? TestML::Num->new(value => $value + 0) :
        "$value" eq "$TestML::Constant::True" ? $value :
        "$value" eq "$TestML::Constant::False" ? $value :
        "$value" eq "$TestML::Constant::None" ? $value :
        TestML::Str->new(value => $value);
}

sub compile_testml {
    my $self = shift;
    my $path = ref($self->testml)
        ? $self->testml
        : join '/', $self->base, $self->testml;
    my $function = TestML::Compiler->new(base => $self->base)->compile($path)
        or die "TestML document failed to compile";
    return $function;
}

sub load_variables {
    my $self = shift;
    my $global = $self->function->outer;
    $global->setvar(Block => TestML::Block->new);
    $global->setvar(Label => TestML::Str->new(value => '$BlockLabel'));
    $global->setvar(True => $TestML::Constant::True);
    $global->setvar(False => $TestML::Constant::False);
    $global->setvar(None => $TestML::Constant::None);
}

sub load_transform_module {
    my $self = shift;
    my $module = shift;
    if ($module ne 'main') {
        eval "require $module; 1"
            or die "Can't use $module:\n$@";
    }

    my $global = $self->function->outer;
    no strict 'refs';
    for my $key (sort keys %{"$module\::"}) {
        next if $key eq "\x16";
        my $glob = ${"$module\::"}{$key};
        if (my $function = *$glob{CODE}) {
            $global->setvar(
                $key => TestML::Native->new(value => $function),
            );
        }
        elsif (my $object = *$glob{SCALAR}) {
            if (ref($$object)) {
                $global->setvar($key => $$object);
            }
        }
    }
    $self->function;
}

sub get_label {
    my $self = shift;
    my $label = $self->function->getvar('Label')->value;
    sub label {
        my $self = shift;
        my $var = shift;
        my $block = $self->function->getvar('Block');
        return $block->label if $var eq 'BlockLabel';
        if (my $v = $block->points->{$var}) {
            $v =~ s/\n.*//s;
            $v =~ s/^\s*(.*?)\s*$/$1/;
            return $v;
        }
        if (my $v = $self->function->getvar($var)) {
            return $v->value;
        }
    }
    $label =~ s/\$(\w+)/label($self, $1)/ge;
    return $label ? ($label) : ();
}

sub run_plan {
    my $self = shift;
    if (! $self->planned) {
        $self->title();
        $self->plan_begin();
        $self->planned(1);
    }
}

sub get_error {
    my $self = shift;
    return $self->function->expression->error;
}

sub clear_error {
    my $self = shift;
    return $self->function->expression->error(undef);
}

sub throw {
    require Carp;
    Carp::croak $_[1];
}

#-----------------------------------------------------------------------------
package TestML::Function;
use TestML::Base -base;

has 'type' => 'Func';       # Functions are TestML typed objects
# XXX Make this a featherweight reference.
has 'signature' => [];      # Input variable names
has 'namespace' => {};      # Lexical scoped variable stash
has 'statements' => [];     # Exexcutable code statements
has 'data' => [];           # Data section scoped to this function

# Runtime pointers to current objects.
has 'expression';
has 'block';

my $outer = {};
sub outer { @_ == 1 ? $outer->{$_[0]} : ($outer->{$_[0]} = $_[1]) }

sub getvar {
    my $self = shift;
    my $name = shift;
    while ($self) {
        if (my $object = $self->namespace->{$name}) {
            return $object;
        }
        $self = $self->outer;
    }
    return;
}

sub setvar {
    my $self = shift;
    my $name = shift;
    my $object = shift;
    $self->namespace->{$name} = $object;
    return;
}

sub forgetvar {
    my $self = shift;
    my $name = shift;
    delete $self->namespace->{$name};
    return;
}

#-----------------------------------------------------------------------------
package TestML::Statement;
use TestML::Base -base;

has 'expression', -init => 'TestML::Expression->new';
has 'assertion';
has 'points' => [];

#-----------------------------------------------------------------------------
package TestML::Expression;
use TestML::Base -base;

has 'units' => [];
has 'error';

#-----------------------------------------------------------------------------
package TestML::Assertion;
use TestML::Base -base;

has 'name';
has 'expression', -init => 'TestML::Expression->new';

#-----------------------------------------------------------------------------
package TestML::Transform;
use TestML::Base -base;

has 'name';
has 'args' => [];
has 'explicit_call' => 0;

#-----------------------------------------------------------------------------
package TestML::Block;
use TestML::Base -base;

has 'label' => '';
has 'points' => {};

#-----------------------------------------------------------------------------
package TestML::Object;
use TestML::Base -base;

has 'value';

sub type {
    my $type = ref(shift);
    $type =~ s/^TestML::// or die "Can't find type of '$type'";
    return $type;
}

sub runtime { return $TestML::Runtime::self }

sub str { my $t = $_[0]->type; die "Cast from $t to Str is not supported" }
sub num { my $t = $_[0]->type; die "Cast from $t to Num is not supported" }
sub bool { my $t = $_[0]->type; die "Cast from $t to Bool is not supported" }
sub list { my $t = $_[0]->type; die "Cast from $t to List is not supported" }
sub none { $TestML::Constant::None }

#-----------------------------------------------------------------------------
package TestML::Str;
use TestML::Object -base;

sub str { shift }
sub num { TestML::Num->new(
    value => ($_[0]->value =~ /^-?\d+(?:\.\d+)$/ ? ($_[0]->value + 0) : 0),
)}
sub bool {
    length($_[0]->value) ? $TestML::Constant::True : $TestML::Constant::False
}
sub list { TestML::List->new(value => [split //, $_[0]->value]) }

#-----------------------------------------------------------------------------
package TestML::Num;
use TestML::Object -base;

sub str { TestML::Str->new(value => $_[0]->value . "") }
sub num { shift }
sub bool { ($_[0]->value != 0) ? $TestML::Constant::True : $TestML::Constant::False }
sub list {
    my $list = [];
    $#{$list} = int($_[0]) -1;
    TestML::List->new(value =>$list);
}

#-----------------------------------------------------------------------------
package TestML::Bool;
use TestML::Object -base;

sub str { TestML::Str->new(value => $_[0]->value ? "1" : "") }
sub num { TestML::Num->new(value => $_[0]->value ? 1 : 0) }
sub bool { shift }

#-----------------------------------------------------------------------------
package TestML::List;
use TestML::Object -base;
sub list { shift }

#-----------------------------------------------------------------------------
package TestML::None;
use TestML::Object -base;

sub str { Str('') }
sub num { Num(0) }
sub bool { $TestML::Constant::False }
sub list { List([]) }

#-----------------------------------------------------------------------------
package TestML::Error;
use TestML::Object -base;

#-----------------------------------------------------------------------------
package TestML::Native;
use TestML::Object -base;

package TestML::Constant;

our $True = TestML::Bool->new(value => 1);
our $False = TestML::Bool->new(value => 0);
our $None = TestML::None->new;



inc/TestML/Compiler.pm
#line 1
package TestML::Compiler;

use TestML::Base -base;
use TestML::Grammar;

has 'base';
has 'debug' => '0';

sub compile {
    my $self = shift;
    my $file = shift;
    if (not ref $file and $file !~ /\n/) {
        $file =~ s/(.*)\/(.*)/$2/ or die;
        $self->base($1);
    }
    my $input = (not ref($file) and $file =~ /\n/)
        ? $file
        : $self->slurp($file);

    my $result = $self->preprocess($input, 'top');

    my ($code, $data) = @$result{qw(code data)};

    my $debug = $self->debug;
    $debug = $result->{DebugPegex} if defined $result->{DebugPegex};
    my $grammar = TestML::Grammar->new(
        receiver => TestML::Receiver->new,
        debug => $debug,
    );
    $grammar->parse($code, 'code_section')
        or die "Parse TestML code section failed";

    $self->fixup_grammar($grammar, $result);

    $grammar->parse($data, 'data_section')
        or die "Parse TestML data section failed";

    if ($result->{DumpAST}) {
        XXX($grammar->receiver->function);
    }

    my $function = $grammar->receiver->function;
    $function->outer(TestML::Function->new());

    return $function;
}

sub preprocess {
    my $self = shift;
    my $text = shift;
    my $top = shift;

    my @parts = split /^((?:\%\w+.*|\#.*|\ *)\n)/m, $text;

    $text = '';

    my $result = {
        TestML => '',
        DataMarker => '',
        BlockMarker => '===',
        PointMarker => '---',
    };

    my $order_error = 0;
    for my $part (@parts) {
        next unless length($part);
        if ($part =~ /^(\#.*|\ *)\n/) {
            $text .= "\n";
            next;
        }
        if ($part =~ /^%(\w+)\s*(.*?)\s*\n/) {
            my ($directive, $value) = ($1, $2);
            $text .= "\n";
            if ($directive eq 'TestML') {
                die "Invalid TestML directive"
                    unless $value =~ /^\d+\.\d+$/;
                die "More than one TestML directive found"
                    if $result->{TestML};
                $result->{TestML} = TestML::Str->new(value => $value);
                next;
            }
            $order_error = 1 unless $result->{TestML};
            if ($directive eq 'Include') {
                $text .= $self->preprocess($self->slurp($value))->{text};
            }
            elsif ($directive =~ /^(DataMarker|BlockMarker|PointMarker)$/) {
                $result->{$directive} = $value;
            }
            elsif ($directive =~ /^(DebugPegex|DumpAST)$/) {
                $value = 1 unless length($value);
                $result->{$directive} = $value;
            }
            else {
                die "Unknown TestML directive '$directive'";
            }
        }
        else {
            $order_error = 1 if $text and not $result->{TestML};
            $text .= $part;
        }
    }

    if ($top) {
        die "No TestML directive found"
            unless $result->{TestML};
        die "%TestML directive must be the first (non-comment) statement"
            if $order_error;

        my $DataMarker = $result->{DataMarker} ||= $result->{BlockMarker};
        my ($code, $data);
        if ((my $split = index($text, "\n$DataMarker")) >= 0) {
            $result->{code} = substr($text, 0, $split + 1);
            $result->{data} = substr($text, $split + 1);
        }
        else {
            $result->{code} = $text;
            $result->{data} = '';
        }

        $result->{code} =~ s/^\\(\\*[\%\#])/$1/gm;
        $result->{data} =~ s/^\\(\\*[\%\#])/$1/gm;
    }
    else {
        $result->{text} = $text;
    }

    return $result;
}

sub fixup_grammar {
    my $self = shift;
    my $grammar = shift;
    my $hash = shift;

    my $namespace = $grammar->receiver->function->namespace;
    $namespace->{TestML} = $hash->{TestML};

    $grammar = $grammar->grammar;
    my $point_lines = $grammar->{point_lines}{'+re'};

    my $block_marker = $hash->{BlockMarker};
    if ($block_marker) {
        $block_marker =~ s/([\$\%\^\*\+\?\|])/\\$1/g;
        $grammar->{block_marker}{'+re'} = qr/\G$block_marker/;
        $point_lines =~ s/===/$block_marker/;
    }

    my $point_marker = $hash->{PointMarker};
    if ($point_marker) {
        $point_marker =~ s/([\$\%\^\*\+\?\|])/\\$1/g;
        $grammar->{point_marker}{'+re'} = qr/\G$point_marker/;
        $point_lines =~ s/---/$point_marker/;
    }

    $grammar->{point_lines}{'+re'} = qr/$point_lines/;
}

sub slurp {
    my $self = shift;
    my $file = shift;
    my $fh;
    if (ref($file)) {
        $fh = $file;
    }
    else {
        my $path = join '/', $self->base, $file;
        open $fh, $path
            or die "Can't open '$path' for input: $!";
    }
    local $/;
    return <$fh>;
}


#-----------------------------------------------------------------------------
package TestML::Receiver;
use TestML::Base -base;

use TestML::Runtime;

has 'function', -init => 'TestML::Function->new()';

has 'stack' => [];
has 'block';

has 'string';
has 'point_name';

my %ESCAPES = (
    '\\' => '\\',
    "'" => "'",
    'n' => "\n",
    't' => "\t",
    '0' => "\0",
);

sub got_single_quoted_string {
    my $self = shift;
    my $string = shift;
    $string =~ s/\\([\\\'])/$ESCAPES{$1}/g;
    $self->string($string);
}

sub got_double_quoted_string {
    my $self = shift;
    my $string = shift;
    $string =~ s/\\([\\\"nt])/$ESCAPES{$1}/g;
    $self->string($string);
}

sub got_unquoted_string {
    my $self = shift;
    $self->string(shift);
}

sub try_assignment_statement {
    my $self = shift;
    push @{$self->function->statements}, TestML::Statement->new();
    my $expression = $self->function->statements->[-1]->expression;
    $expression->units->[0] = TestML::Transform->new(name => 'Set');
    $self->function->statements->[-1]->expression($expression);
    push @{$self->stack}, TestML::Expression->new;
}

sub got_assignment_statement {
    my $self = shift;
    $self->function->statements->[-1]->expression->units->[0]->args->[1] =
        pop @{$self->stack};
}

sub not_assignment_statement {
    my $self = shift;
    pop @{$self->function->statements};
    pop @{$self->stack};
}

sub got_variable_name {
    my $self = shift;
    my $variable_name = shift;
    $self->function->statements->[-1]->expression->units->[0]->args->[0] =
        $variable_name;
}

sub try_code_statement {
    my $self = shift;
    push @{$self->function->statements}, TestML::Statement->new();
    push @{$self->stack}, $self->function->statements->[-1]->expression;
}

sub got_code_statement {
    my $self = shift;
    pop @{$self->stack};
}

sub not_code_statement {
    my $self = shift;
    pop @{$self->function->statements};
    pop @{$self->stack};
}

sub got_point_object {
    my $self = shift;
    my $point_name = shift;
    $point_name =~ s/^\*// or die;
    push @{$self->stack->[-1]->units},
        TestML::Transform->new(
            name => 'Point',
            args => [$point_name],
        );
    push @{$self->function->statements->[-1]->points}, $point_name;
}

sub try_function_object {
    my $self = shift;
    my $function = TestML::Function->new(outer => $self->function);
    $self->function($function);
}

sub got_function_object {
    my $self = shift;
    push @{$self->stack->[-1]->units}, $self->function;
    $self->function($self->function->outer);
}

sub not_function_object {
    my $self = shift;
    $self->function($self->function->outer);
}

sub got_function_variable {
    my $self = shift;
    push @{$self->function->signature}, shift;
}

sub try_transform_object {
    my $self = shift;
    push @{$self->stack->[-1]->units}, TestML::Transform->new();
}

sub not_transform_object {
    my $self = shift;
    pop @{$self->stack->[-1]->units};
}

sub got_transform_name {
    my $self = shift;
    $self->stack->[-1]->units->[-1]->name(shift);
}

sub got_transform_argument_list {
    my $self = shift;
    $self->stack->[-1]->units->[-1]->explicit_call(1);
}

sub try_transform_argument {
    my $self = shift;
    push @{$self->stack}, TestML::Expression->new;
}

sub got_transform_argument {
    my $self = shift;
    my $argument = pop @{$self->stack};
    push @{$self->stack->[-1]->units->[-1]->args}, $argument;
}

sub not_transform_argument {
    my $self = shift;
    pop @{$self->stack};
}

sub got_string_object {
    my $self = shift;
    push @{$self->stack->[-1]->units},
        TestML::Str->new(value => $self->string);
}

sub got_number_object {
    my $self = shift;
    push @{$self->stack->[-1]->units},
        TestML::Num->new(value => shift);
}

sub try_assertion_call {
    my $self = shift;
    $self->function->statements->[-1]->assertion(TestML::Assertion->new);
    push @{$self->stack},
        $self->function->statements->[-1]->assertion->expression;
}

sub got_assertion_call {
    my $self = shift;
    pop @{$self->stack};
}

sub not_assertion_call {
    my $self = shift;
    $self->function->statements->[-1]->assertion(undef);
    pop @{$self->stack};
}

sub got_assertion_eq {
    my $self = shift;
    $self->function->statements->[-1]->assertion->name('EQ');
}

sub got_assertion_ok {
    my $self = shift;
    $self->function->statements->[-1]->assertion->name('OK');
}

sub got_assertion_has {
    my $self = shift;
    $self->function->statements->[-1]->assertion->name('HAS');
}

sub got_block_label {
    my $self = shift;
    $self->block(TestML::Block->new(label => shift));
}

sub got_point_name {
    my $self = shift;
    $self->point_name(shift);
}

sub got_point_phrase {
    my $self = shift;
    my $point_phrase = shift;
    $self->block->points->{$self->point_name} = $point_phrase;
}

sub got_point_lines {
    my $self = shift;
    my $point_lines = shift;
    $self->block->points->{$self->point_name} = $point_lines;
}

sub got_data_block {
    my $self = shift;
    push @{$self->function->data}, $self->block;
}


inc/Module/Install/VersionCheck.pm
#line 1
package Module::Install::VersionCheck;
use strict;
use warnings;
use 5.008003;

use Module::Install::Base;

my $DEFAULT = '0.00';

use vars qw($VERSION @ISA);
BEGIN {
    $VERSION = '0.10';
    @ISA     = 'Module::Install::Base';
}

sub version_check {
    my $self = shift;
    return unless $self->is_admin;

    my $module_version = $self->_get_module_version();
    my $changes_version = $self->_get_changes_version();
    my $git_tag_version = $self->_get_git_tag_version();

    $self->_report(
        $module_version,
        $changes_version,
        $git_tag_version,
    );
}

sub _get_module_version {
    my $self = shift;
    return $DEFAULT unless $self->admin->{extensions};
    my ($metadata) = grep {
        ref($_) eq 'Module::Install::Metadata';
    } @{$self->admin->{extensions}};
    return $DEFAULT unless $metadata;
    return $metadata->{values}{version} || $DEFAULT;
}

sub _get_changes_version {
    my $self = shift;
    return $DEFAULT unless -e 'Changes';
    open IN, 'Changes' or die "Can't open 'Changes' for input: $!";
    my $text = do {local $/; <IN>};
    $text =~ /\b(\d\.\d\d)\b/ or return $DEFAULT;
    return $1;
}

sub _get_git_tag_version {
    my $self = shift;
    return $DEFAULT unless -e '.git';
    require Capture::Tiny;
    my $text = Capture::Tiny::capture_merged(sub { system('git tag') });
    my $version = $DEFAULT;
    for (split "\n", $text) {
        if (/\b(\d\.\d\d)\b/ and $1 > $version) {
            $version = $1;
        }
    }
    return $version;
}

sub _report {
    my $self = shift;
    print "version_check found: @_\n";
}

1;

=encoding utf8

#line 107


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.019.
use strict;
use warnings;

use 5.008001;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Get a hash of your current 'my' variables",
  "AUTHOR" => "Ingy d\x{f6}t Net <ingy\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "App-AYCABTU",
  "EXE_FILES" => [
    "bin/aycabtu"
  ],
  "LICENSE" => "perl",
  "NAME" => "App::AYCABTU",
  "PREREQ_PM" => {
    "Capture::Tiny" => "0.10",
    "Mouse" => "0.93",
    "YAML::XS" => "0.35"
  },
  "VERSION" => "0.12",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Capture::Tiny" => "0.10",
  "Mouse" => "0.93",
  "YAML::XS" => "0.35"
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



